tests/NavigationTests.elm

```diff
@@ -0,0 +1,156 @@
+initWithStringFlags : String -> Navigation.Location -> ( Model, Cmd Msg )
+initWithStringFlags flags location =
+    init location
+
+
+programUpdate : Msg -> Model -> ( Model, Cmd Msg )
+programUpdate msg model =
+    case msg of
+        UrlChange location ->
+            ( { model | location = location, msgs = msg :: model.msgs }, Navigation.modifyUrl location.pathname )
+
+        PushUrl url ->
+            ( model, Navigation.newUrl url )
+
+        ModifyUrl url ->
+            ( model, Navigation.modifyUrl url )
+
+        Back amount ->
+            ( model, Navigation.back amount )
+
+        Forward amount ->
+            ( model, Navigation.forward amount )
+
+        Load url ->
+            ( model, Navigation.load url )
+
+        Reload ->
+            ( model, Navigation.reload )
+
+
+sampleProgram : TestContext Model Msg
+sampleProgram =
+    Navigation.program UrlChange
+        { init = init
+        , update = programUpdate
+        , subscriptions = \_ -> Sub.none
+        , view = always <| div [] []
+        }
+        |> TestContext.start
+
+
+sampleProgramWithFlags : TestContext Model Msg
+sampleProgramWithFlags =
+    Navigation.programWithFlags UrlChange
+        { init = initWithStringFlags
+        , update = programUpdate
+        , subscriptions = \_ -> Sub.none
+        , view = always <| div [] []
+        }
+        |> TestContext.startWithFlags "foo"
+
+
+all : Test
+all =
+    describe "Navigation TestContext"
+        [ test "starting a navigation program" <|
+            \() ->
+                sampleProgram
+                    |> expectHref "https://elm.testable/"
+        , test "pushUrl" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "modifyUrl" <|
+            \() ->
+                sampleProgram
+                    |> update (ModifyUrl "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "back" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/foo")
+                    |> update (PushUrl "/bar")
+                    |> update (Back 1)
+                    |> expectHref "https://elm.testable/foo"
+        , test "forward" <|
+            \() ->
+                sampleProgram
+                    |> update (PushUrl "/bar")
+                    |> update (PushUrl "/baz")
+                    |> update (Back 2)
+                    |> update (Forward 1)
+                    |> expectHref "https://elm.testable/bar"
+        , test "load" <|
+            \() ->
+                sampleProgram
+                    |> update (Load "/foo")
+                    |> expectHref "https://elm.testable/foo"
+        , test "refresh does nothing" <|
+            \() ->
+                sampleProgram
+                    |> update Reload
+                    |> done
+        , describe "navigation simulation"
```

Comment:
Yes, here I've tried to do more higher level tests, testing things integrated, with `testable/NavigationTests` on a more unitary level, also to make it easier to extract the location encoding/parsing to a lib later